### PR TITLE
implement NodeConfig merging for ec2 user-data

### PR DESCRIPTION
**Pull request by ndbaker1**
**Issue #, if available:**

**Description of changes:**

implement merge semantics for `NodeConfig`, which originally returned an error when encountering more than one `NodeConfig` in a userdata MIME multipart document.

* the blanket merge strategy for the `NodeConfig` is fields in the **patch/source** document will overwrite corresponding fields in the **base/destination** document.
* the merge strategy for `Spec.Kubelet.Config`, the inline kubelet config, is a full json-merge between the two unless one is empty. In the latter case, use the content from the non-empty field.
* the merge strategy for `Spec.Kubelet.Flags`, the kubelet flag arguments, is to append the source flags to the destination flags. Following the semantics of the underlying flag parsing library, https://github.com/spf13/pflag, single-value flags that come last have higher precedence, and multiple-value flags are merged with the library's underlying strategy (which obeys user expectations rather than us using some heuristic)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Testing Done**

*[See this guide for recommended testing for PRs.](../doc/CONTRIBUTING.md#testing-changes) Some tests may not apply. Completing tests and providing additional validation steps are not required, but it is recommended and may reduce review time and time to merge.*

### CODE REVIEW

File: nodeadm/internal/api/merge.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+			if err := json.Unmarshal([]byte(src.String()), &srcConfigMap); err != nil {
+				return err
+			}
+			if err := mergo.Merge(&dstConfigMap, &srcConfigMap, mergo.WithOverride); err != nil {
+				return err
+			}
+			config, err := json.MarshalIndent(dstConfigMap, "", "    ")
+			if err != nil {
+				return err
+			}
```
Comment (cartermckinnon) [inline nodeadm/internal/api/merge.go:78]: This looks really similar to the merging logic in #1612, lets make sure we can share the impl

File: nodeadm/internal/configprovider/userdata.go [hunk 2]
```diff
@@ -83,23 +88,27 @@ func parseMultipart(userDataReader *multipart.Reader) (*internalapi.NodeConfig,
...
 				if err != nil {
 					return nil, err
 				}
+				if config != nil {
+					if err := config.Merge(decodedConfig); err != nil {
+						return nil, err
+					}
+				} else {
+					config = decodedConfig
+				}
```
Comment (cartermckinnon) [inline nodeadm/internal/configprovider/userdata.go:111]: I think it'd be a bit clearer to just collect the decoded configs in a slice and then merge them all at the end.

**Review by cartermckinnon — CHANGES_REQUESTED** [review]
Couple small things

File: nodeadm/internal/api/merge.go (hunk 1 above)
Comment (ndbaker1) [reply nodeadm/internal/api/merge.go]: put this into a `DocumentMerge` util, which basically uses marshalled representations of `any`-types to the do underlying merge. not beautiful but its not like we're we need to hunt for optimizations just yet 🤷

File: nodeadm/internal/configprovider/userdata.go (hunk 2 above)
Comment (ndbaker1) [reply nodeadm/internal/configprovider/userdata.go]: done!

**Review by cartermckinnon — APPROVED** [review]

**Pull request merged**